Adding nvidia-hpc-sdk based on upstream PR

pramodk opened the pull request:
- based on new PR upstream https://github.com/spack/spack/pull/19139

pramodk committed:
Adding nvidia-hpc-sdk based on upstream PR
  - new PR upstream https://github.com/spack/spack/pull/19139

pramodk commented:
retest this please

matz-e commented on var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:
```diff
@@ -0,0 +1,116 @@
+    maintainers = ['Nischay-Pro']
+
+    for ver, packages in _versions.items():
+        key = "{0}_{1}".format(platform.system(), platform.machine())
```
So… no cross-compiling :)

pramodk replied:
yes, no cross compiling but you just have to by the node with target arch :D

samcmill committed:
New compiler: nvhpc (NVIDIA HPC SDK) (#19294)

* Add nvhpc compiler definition: "spack compiler add" will now look
  for instances of the NVIDIA HPC SDK compiler executables
  (nvc, nvc++, nvfortran) in supplied paths
* Add the nvhpc package which installs the nvhpc compiler
* Add testing for nvhpc detection and C++-standard/pic flags

Co-authored-by: Scott McMillan <smcmillan@nvidia.com>

Based on https://github.com/spack/spack/pull/19294

pramodk committed:
Add nvidia-hpc-sdk to whitelist

pramodk commented:
Retest this please

pramodk commented:
Retest this please

matz-e commented on lib/spack/spack/compilers/nvhpc.py:
```diff
@@ -0,0 +1,90 @@
+from spack.compiler import Compiler
+
+
+class Nvhpc(Compiler):
```
Is this from upstream?

pramodk replied:
yup!

matz-e reviewed (dismissed):
Seems good to me, I guess the hold-up will be MPI?

pramodk committed:
Add CUDA@11 required for latest NVIDIA-HPC-SDK

ohm314 reviewed (dismissed):
lgtm

pramodk commented:
> Seems good to me, I guess the hold-up will be MPI?

I assume you are referring to `-isystem` flag. We are using FindMPI in CoreNEURON and not depend on mpicc/mpicxx wrappers. So that allows us to continue using PGI.

pramodk committed:
Fix legacy apis : setup_environment to setup_run_environment
